### ISSUE TITLE
C++11?

### ISSUE THREAD

**Issue by vinniefalco** [jk-jeon/dragonbox#8]
I need a port of this to C++11

**Comment by expnkx** [comment]
please. let C++11 die. thank you

**Comment by jk-jeon** [comment]
It'll be a lot of work. Since the new quarter has begun, I can't afford much time to it, but I'll try.

**Comment by expnkx** [comment]
> It'll be a lot of work. Since the new quarter has begun, I can't afford much time to it, but I'll try.

don't waste your time on C++11 tbh. C++11 should die. It is terrible

**Comment by stainleebakhla** [comment]
Would love it if C++11 version of this is made available. Great work guys!

**Comment by ecorm** [comment]
I'm sure everyone here already knows this, but I'll mention it for those who don't: `constexpr if` can be emulated in C++11 by tagged dispatching using `std::true_type` and `std::false_type`:

Instead of
```c++
if constexpr(condition)
{
    do_this(arg1);
}
else
{
    do_that(arg2);
}
```

one can do:

```c++
do_this_or_that(std::integral_constant<bool, condition>{}, arg1, arg2);

void do_this_or_that(std::true_type, int arg1, int arg2)
{
    do_this(arg1);
}

void do_this_or_that(std::false_type, int arg1, int arg2)
{
    do_that(arg2);
}
```

Any decent compiler will optimize this into the equivalent `if constexpr` code.

The downside of course is more code, and the need to pass variables as arguments. The upside is compatibility with C++11 and (presumably) wider adoption.

`std::bool_constant` doesn't exist prior to C++17, but one can easily define their own template alias that does the same thing:

```c++
template <bool condition>
using meta_bool = std::integral_constant<bool, condition>;
```

then
```c++
do_this_or_that(meta_bool<condition>{}, arg1, arg2);
```

**Comment by ecorm** [comment]
According to this [2020 survey](https://blog.jetbrains.com/clion/2020/06/dev-eco-cpp-2020/), 46% responded that they regularly use C++11, so it's far from being dead.

I can understand wanting to use the latest C++ features for writing an _application_, but for libraries, IMHO, it's better to be conservative in the minimum C++ version required if you want the widest possible adoption.

**Comment by vinniefalco** [comment]
> if you want the widest possible adoption.

GitHub stars and forks are what give my life meaning and value

**Comment by ecorm** [comment]
> GitHub stars and forks are what give my life meaning and value

I'm working on a serialization library intended to be used in conjunction with Asio and your Beast library. It's currently closed source, but I may be able to convince my employer to make it open source in the future. I'm deliberately constraining my serialization library to C++11 so that its language requirements matches that of yours. I would love to make use of Dragonbox for fast float-to-string conversions, but it would then artificially bump the requirements of my library up to C++17.

**Comment by vinniefalco** [comment]
>  I would love to make use of Dragonbox for fast float-to-string conversions

I agree that requiring C++17 will definitely reduce the size of the potential audience. If you want, you could just crib the RYU algorithm from Boost.JSON which perform float-to-string conversions. It is comparable in terms of performance and requires only C++11:
https://github.com/boostorg/json/tree/develop/include/boost/json/detail/ryu

**Comment by expnkx** [comment]
> I'm sure everyone here already knows this, but I'll mention it for those who don't: `constexpr if` can be emulated in C++11 by tagged dispatching using `std::true_type` and `std::false_type`:
> 
> Instead of
> 
> ```c++
> if constexpr(condition)
> {
>     do_this(arg1);
> }
> else
> {
>     do_that(arg2);
> }
> ```
> 
> one can do:
> 
> ```c++
> do_this_or_that(std::integral_constant<bool, condition>{}, arg1, arg2);
> 
> void do_this_or_that(std::true_type, int arg1, int arg2)
> {
>     do_this(arg1);
> }
> 
> void do_this_or_that(std::false_type, int arg1, int arg2)
> {
>     do_that(arg2);
> }
> ```
> 
> Any decent compiler will optimize this into the equivalent `if constexpr` code.
> 
> The downside of course is more code, and the need to pass variables as arguments. The upside is compatibility with C++11 and (presumably) wider adoption.
> 
> `std::bool_constant` doesn't exist prior to C++17, but one can easily define their own template alias that does the same thing:
> 
> ```c++
> template <bool condition>
> using meta_bool = std::integral_constant<bool, condition>;
> ```
> 
> then
> 
> ```c++
> do_this_or_that(meta_bool<condition>{}, arg1, arg2);
> ```

that is not zero-overhead tbh. On Microsoft ABI, even empty objects would waste registers.

**Comment by vinniefalco** [comment]
> I failed to understand why you cannot build the latest toolchain/cross-toolchain. It is extremely simple okay?

If this was universally true, you wouldn't see so many projects which still use C++11.

**Comment by expnkx** [comment]
> > I failed to understand why you cannot build the latest toolchain/cross-toolchain. It is extremely simple okay?
> 
> If this was universally true, you wouldn't see so many projects which still use C++11.

That is universally true. It is extremely simple.

**Comment by expnkx** [comment]
![image](https://user-images.githubusercontent.com/57303909/112740260-8283ea80-8f49-11eb-9005-938b6d0d2c15.png)
Funny you are talking about compatibility and portable without knowing how to build cross toolchains and test on them.

**Comment by vinniefalco** [comment]
> That is universally true. 

Then why are there so many projects which use C++11? From https://www.jetbrains.com/lp/devecosystem-2020/cpp/

![image](https://user-images.githubusercontent.com/1503976/112740281-82c3bc00-8f30-11eb-895a-0b464d556549.png)

Why are 46% of projects using only C++11? If it is "extremely simple?"

**Comment by expnkx** [comment]
> > That is universally true.
> 
> Then why are there so many projects which use C++11? From https://www.jetbrains.com/lp/devecosystem-2020/cpp/
> 
> ![image](https://user-images.githubusercontent.com/1503976/112740281-82c3bc00-8f30-11eb-895a-0b464d556549.png)
> 
> Why are 46% of projects using only C++11? If it is "extremely simple?"

1. Why? it is extremely simple. C++ is bad because of people like you who dumb it down like abusing unique_ptr.
2. You truly believe this graph? People around me are stick to C++98 dude.

**Comment by ecorm** [comment]
Let's keep the discussion civil.

> that is not zero-overhead tbh. On Microsoft ABI, even empty objects would waste registers

If the code is all inline, wouldn't the compiler optimize the function call away, regardless of ABI?

> I failed to understand why you cannot build the latest toolchain/cross-toolchain. It is extremely simple okay? Even on windows.

It's not me you need to convince, but the 46% of my library's potential audience who still use C++11 (see survey link in my above post). If it were a in-house project, then I could of course use the latest toolchains if I so choose.

Another thing to keep in mind is that many open source projects are included in Linux distros, so they are constrained to the "standard" compiler versions used in those distros.

To be fair, I can understand the point of view that new libraries adopting C++17 helps to the encourage the industry in moving on past C++11. I guess it's up to the author to decide what's important for them: wider adoption or the joy/convenience of using the latest C++ language features.

**Comment by expnkx** [comment]
> If the code is all inline, wouldn't the compiler optimize the function call away, regardless of ABI?

inline in C++ means prevent ODR violation. It has nothing to do with what you are doing. What if that expansion fails? Then you stick with overhead.

**Comment by expnkx** [comment]
it is a vicious cycle. Boost people like you who keep supporting C++11 nonsense and people stuck with you.

Of course there is a reason why boost nowadays is so horrible because they ruined C++.

**Comment by vinniefalco** [comment]
> If the code is all inline, wouldn't the compiler optimize the function call away, regardless of ABI?

Yes, the compiler sees right through it with full optimizations turned on. And if it doesn't, that's a problem with the compiler which should be fixed there as well.

**Comment by expnkx** [comment]
> > If the code is all inline, wouldn't the compiler optimize the function call away, regardless of ABI?
> 
> Yes, the compiler sees right through it with full optimizations turned on. And if it doesn't, that's a problem with the compiler which should be fixed there as well.

Fixed? That is completely false.

The real truth is that the compiler rarely does the right thing. We would assume the compiler does the right thing unless we prove it does not. In fact, it works the opposite, the compiler always does the wrong thing unless we prove it does the right thing.

You do not know how your abstraction is going to hurt compiler optimizations in some form.

**Comment by expnkx** [comment]
You want me to say some sad truth? You need to support C89 dude if you want your stupid standard to be portable. Not using any C++ trash.

**Comment by expnkx** [comment]
```cpp
void do_this_or_that(std::true_type, int arg1, int arg2)
{
    do_this(arg1);
}

void do_this_or_that(std::false_type, int arg1, int arg2)
{
    do_that(arg2);
}

```

let me be clear. you already invoke overhead here. because you do not mark them as inline, they are now dead code now.

**Comment by expnkx** [comment]
![image](https://user-images.githubusercontent.com/57303909/112740639-7b121080-8f4c-11eb-8e07-c84607a63b69.png)
see? dead code

**Comment by ecorm** [comment]
Then just add the `inline` that I forgot.

**Comment by expnkx** [comment]
> Then just add the `inline` that I forgot.

things you forgot or do not understand than what you understand.

**Comment by vinniefalco** [comment]
It seems there are diminishing returns (really, no returns) on additional dialogue with this individual.

**Comment by expnkx** [comment]
> It seems there are diminishing returns (really, no returns) on additional dialogue with this individual.

You want performance but you do not update your compiler lol.

**Comment by ecorm** [comment]
@vinniefalco  He got kicked out of the Reddit C++ subforum, as well as the Ryu Github repo. Not hard to imagine why.

**Comment by expnkx** [comment]
> @vinniefalco He got kicked out of the Reddit C++ subforum, as well as the Ryu Github repo. Not hard to imagine why.

Reddit C++ subforum? LOL.

I am now the super moderator of C++ tieba in China which contains 4x more people than reddit C++ subforum trash. You are OO losers.
Who cares about reddit C++ subforum. I am 4x more powerful than entire C++ subforum.

**Comment by expnkx** [comment]
![image](https://user-images.githubusercontent.com/57303909/112740792-e14b6300-8f4d-11eb-9e60-668bd3dfdc17.png)

**Comment by expnkx** [comment]
Make your own library.

**Comment by expnkx** [comment]
How stupid this issue is?

Do you talk about standards? Then why do you not write C89 code? it is portable EVEN for other languages like Rust or python to use them.

Do you talk about portability? You do not even understand how to build cross toolchains and test them on them. 

Why bitch about things when you can make your own library? Do you pay JK money for doing so?

Why updating compilers is so hard for you?

Boost? Remember the boost regex, boost filesystem disasters? How do those things work out in C++ standard? There is a reason why boost is such a horrible library right now.

Oh, you talk about Reddit? I am 4x more powerful than your Reddit moderators technically.

**Comment by ecorm** [comment]
You don't pay the author either. We're simply making a feature request and presented some evidence on the number of C++11 users. I'm trying to stay civil here, but you turned this into a angry rant. Even if you're technically correct, people won't take you seriously if you don't remain calm and respectful.

If you admire Linus Torvalds so much, perhaps you should do like him and try to learn some empathy: https://arstechnica.com/gadgets/2018/09/linus-torvalds-apologizes-for-years-of-being-a-jerk-takes-time-off-to-learn-empathy/

Of course if the author declines this C++11 feature request, I'll respect that decision and will use another alternative, such as Ryu, or forking this project.

I have said all I have to say about this subject. Good day to you.

**Comment by expnkx** [comment]
> You don't pay the author either. We're simply making a feature request and presented some evidence on the number of C++11 users. I'm trying to stay civil here, but you turned this into a angry rant. Even if you're technically correct, people won't take you seriously if you don't remain calm and respectful.
> 
> If you admire Linus Torvalds so much, perhaps you should do like him and try to learn some empathy: https://arstechnica.com/gadgets/2018/09/linus-torvalds-apologizes-for-years-of-being-a-jerk-takes-time-off-to-learn-empathy/
> 
> Of course if the author declines this C++11 feature request, I'll respect that decision and will use another alternative, such as Ryu, or forking this project.
> 
> I have said all I have to say about this subject. Good day to you.

I do not believe C++11 is the mainstream C++ usage at all because in my tieba (which is like a forum), every day people are asking where to download VC++6.0 or dev-c++4.9.9.2. They are all C++89 compiler. The statistics of ISO C++ is already seriously flawed since they only investigate people who can access the ISO C++ official website. Majority of people are still stuck in C++89.

Not even mentioning there are a huge amount of people across the globe who do not speak English and do not join the investigation.

I know people are hating me because I am speaking the sad truth. Do you know how bad the current situation is for C++? Why do we need an external library like dragonbox or ryu just for printing floating points? Why? because iostream sucks.

Technically, they are all iostream dialects. they just do want iostream is doing while can never be efficient and composite with each other.

Think about it. C++17 filesystem, charconv, C++20 format, C++23 networking etc are all doing what iostream is doing while creating their own dialects which can never with each other.

**Comment by ecorm** [comment]
> I know people are hating me because I am speaking the sad truth.

There are ways of arguing what you believe is the truth without coming off as a jerk. It's not **what** you say that people dislike, it's the **way** you say it that comes off as insulting. I'm not telling you this to belittle you, but to **help** you become a better person.

If you call other people's work "trash", then it's the same as insulting those people directly. One can critique another person's work without being insulting.

If you can have people respect you for **both** your knowledge and your good manners, then you will truly become an individual with "power".

**Comment by jk-jeon** [comment]
Oh god. What's happening here 😅

@ecorm Thanks for the statistics, the percentage of C++11 looks greater than I expected. It sounds worth to give a try. Right now I do not have time, probably until the beginning of the summer break. Probably I'll work on it after dealing with the items in my personal to-do queue, like coming up with a solution for the issue #11, updating the benchmark, etc..

FYI, @abolz has reimplemented Dragonbox, and AFAIK this implementation does not have lots of compiler-torturing nonsenses mine has, so it is probably C++11 or C++03. He also has an arguably better decimal-to-string conversion routine, which requires a little bit more preconditions but has greater performance. You might be interested to check out the [repository](https://github.com/abolz/Drachennest).

(It was my goal to provide as many customization options as possible because I aimed to provide a library that serves as **a building block for float-to-string conversion library**, not such a library itself. That's one reason for those compiler-torturing nonsenses. Another (possibly-related) reason is that I hate magic numbers and tried to minimize their occurrences. Probably these are just overengineering, common to amateurs😕)

Or, [`fmt`](https://github.com/fmtlib/fmt) should be a viable option as well, which uses Dragonbox internally. Since the version `7.1.0`, it has far better performance compared to the version I tested against in the benchmark. And I believe `fmt` is compatible with C++11. Could be C++14, but I think `fmt` is targeting C++11 and any presence of non-C++11 features should be bugs that you can report.

**Comment by ecorm** [comment]
@jk-jeon Thanks for the information. I was unaware of abolz' implementation. It looks like it's just what I need with Javascript-like behavior for use with JSON.

**Comment by vinniefalco** [comment]
> I aimed to provide a library that serves as a building block for float-to-string conversion library

I discovered on my own that converting a number between base 2 and base 10 is in fact quite difficult :) Thank you for this library.

**Comment by jk-jeon** [comment]
Given that the implementation embedded in `fmt` is already C++11-compatible, re-extracting that into a separate repo should not be a big effort. The ported version will have a much smaller set of configurable policies, but it should not be a big issue to most of the potential users.

But before doing that, I'd like to change some interfaces (and finally release the version1 tag). I'll get back to this after that.

**Comment by robhz786** [comment]
Hi,
I know this thread is about C++11, but in case anyone is interest, I have just adapted the code to C++14, in the branch [cxx14](https://github.com/robhz786/dragonbox/tree/cxx14), in [my fork](https://github.com/robhz786/dragonbox).

It might not be that difficult now to convert it from there to C++11.

**Comment by jk-jeon** [comment]
@robhz786 Oh thank you so much! I'll refer to it when I work on the C++11 port!

**Comment by ecorm** [comment]
Going back to the use of `constexpr if` statements... If the condition is a known constant at compile-time, all decent compilers will optimize-out the branch not taken. Instead of `constexpr if`, you could do something like:

```c++
#if __cplusplus >= 201703L
#define JKJ_CONSTEXPR_IF constexpr if
#else
#define JKJ_CONSTEXPR_IF if
#endif
```

**Comment by ecorm** [comment]
C++ language standards adoption from the [jetbrains report done in 2021](https://blog.jetbrains.com/clion/2021/07/cpp-ecosystem-in-2021/):

![C++ language standards adoption](https://blog.jetbrains.com/wp-content/uploads/2021/07/standards.png)

**Comment by jk-jeon** [comment]
@ecorm Thanks for sharing this!

I'm planning C++11 support (along with constexpr support for C++20) for the next release!

**Comment by ecorm** [comment]
Jetbrains have not published any infographics of their [2022 survey](https://blog.jetbrains.com/blog/2022/04/26/participate-in-the-developer-ecosystem-survey-2022/). 

There is, however, page 19 of the [2022 Annual C++ Developer Survey "Lite"](https://isocpp.org/files/papers/CppDevSurvey-2022-summary.pdf), presumably run by the ISO committee.

Snapshot of relevant chart, which I share here under the _Fair Use Doctrine_ for the purpose of discussion:

![Screenshot from 2023-01-16 17-43-06](https://user-images.githubusercontent.com/9258548/212770070-fee2a16d-3f7c-4256-86ce-9ac3d85ec90d.png)

For C++17, 20% have responded "Partial: Lim...", and around 12% have responded "No: Not allo...". Compare that to ~8% and ~2% respectively for C++11.

Please note that this is not a "nag" to support C++11 for this library. It's just that this issue has become a convenient place for me to keep track of C++11/C++17 usage throughout the years. :grin:

**Comment by alugowski** [comment]
Great chart, @ecorm . It does show that a C++11 version would gain a potential user base of 1/5 of respondents. The previous plots have to be taken with a pinch of salt. Just because someone *is using* C++11 does not mean they *can't*  or *won't* use something newer.

As much as I agree about library compatibility, I think @jk-jeon should feel free to not make it a priority. This is an academic endeavor, after all, and grad school will absolutely not reward the effort (been there). And it must be a lot of effort, as no-one has submitted a PR in 2.5 years. If anything, it would be a net *negative* because it would make further research harder as the code would be more difficult to work with.

If it's stuff like `#define` away the `constexpr`, sure. Otherwise I hope @jk-jeon spends his time on the algorithms and delegate stuff like this to others.

**Comment by jk-jeon** [comment]
@alugowski Thanks so much for your kind words.

>This is an academic endeavor, after all, and grad school will absolutely not reward the effort (been there).

Haha. This project indeed has nothing to do with what I'm supposed to do in my grad school life. I'm really not supposed to waste my time on this but I just can't stop it because it's so fun. At this point I am trying to minimize my time working on this, but I'll never entirely give up on this project.

**Comment by ecorm** [comment]
> It does show that a C++11 version would gain a potential user base of 1/5 of respondents.

If you look at the "No: Not Allowed" column on page 20, the difference between C++11 and C++17 is only around 10%. I'm starting to question my own self-imposition of C++11 in my open-source libraries. It would be nice to get rid of my `string_view` and `optional` polyfills, as well as being able to use `if constexpr`, generic lambdas, auto return types, etc.

**Comment by alugowski** [comment]
> > It does show that a C++11 version would gain a potential user base of 1/5 of respondents.
> 
> If you look at the "No: Not Allowed" column on page 20, the difference between C++11 and C++17 is only around 10%. I'm starting to question my own self-imposition of C++11 in my open-source libraries. It would be nice to get rid of my `string_view` and `optional` polyfills, as well as being able to use `if constexpr`, generic lambdas, auto return types, etc.

You are far more patient than I. One datapoint is what Python extensions can be safely built in. That's been C++17 for quite a while. And most are built in a container with *2014* in its name ;)

**Comment by jk-jeon** [comment]
Features needed to be replaced:
- [ ] Terse `static_assert`
- [ ] `inline constexpr` variables
- [ ] Structured bindings
- [ ] `constexpr` table building
- [ ] `if constexpr`
- [ ] Deduced return type

@mborland Dear Matt, if this doesn't take your precious time too much, could you let me know of anything else you recall?

**Comment by mborland** [comment]
Those are most of the big ones. Lesser issues:

- Lambdas with auto parameters. I replaced those with template functions
- You have the inline constexpr variables, but I found to work with old compilers the cache structs needed to be templated. An example is [here.](https://github.com/cppalliance/charconv/blob/develop/include/boost/charconv/detail/dragonbox/dragonbox.hpp#L490) Since the static variables are implicitly inlined you'll hit linker issues quickly.
- For deduced return types coming from the policies I ended up just hard-coding the policy I wanted in [C++11 mode](https://github.com/cppalliance/charconv/blob/develop/include/boost/charconv/detail/dragonbox/dragonbox.hpp#L2403). I have already added trailing return types to most things if you [save some time](https://github.com/cppalliance/charconv/blob/develop/include/boost/charconv/detail/dragonbox/dragonbox.hpp#L2334)
- I replaced `if constexpr` with `std::enable_if` templates in most places. Dispatching on true/false type would work equally well.
- Most of the `constexpr` functions were as easy as replacing `constexpr` with `BOOST_CONSTEXPR` which just generally just checks the values of the [feature test macros](https://en.cppreference.com/w/cpp/feature_test). 

If you want to reference anything I refactored all of the dragon box and floff code in charconv to be in a folder by itself [here](https://github.com/cppalliance/charconv/tree/develop/include/boost/charconv/detail/dragonbox). I support back to Clang 3.5 and GCC 5. I ended up dropping the GCC 4.X series because of broken aggregate initialization of u128 in the caches. Let me know if you have any additional question and I'll do my best to answer them.

**Comment by jk-jeon** [comment]
@mborland Thanks a lot, it's really helpful!

**Comment by jk-jeon** [comment]
I mostly finished this but I need some rigorous testing. I'll do it shortly but PR is welcomed!

**Comment by jk-jeon** [comment]
Okay, it now at least compiles successfully with:

C++11:
- Clang >= 3.7
- GCC >= 5.1

C++14:
- Clang >= 3.7
- GCC >= 5.1
- MSVC >= 19.14

C++17:
- Clang >= 5.0
- GCC >= 5.1
- MSVC >= 19.14

C++20:
- Clang >= 10.0
- GCC >= 10.1
- MSVC >= 19.30

**Comment by vinniefalco** [comment]
Wow, that's pretty amazing :) Thank you so much. It is true that C++11 is getting pretty old, and my general advice is to target C++ for new libraries. However for something like this, which is such a profound low-level building block that has performance implications, giving people stuck on C++11 access to using the library is a bigger win. Thanks.

**Comment by ecorm** [comment]
ISO's _2024 Annual C++ Developer Survey "Lite"_ is out: https://isocpp.org/files/papers/CppDevSurvey-2024-summary.pdf

![image](https://github.com/jk-jeon/dragonbox/assets/9258548/26f1d1d6-2565-4040-8860-417ba8059b6b)
![image](https://github.com/jk-jeon/dragonbox/assets/9258548/6f0c03af-eb6a-47c0-865f-1cbd8b36b284)

"Pretty much all" for C+11 is 91% versus 78% for C++17.

**Comment by jk-jeon** [comment]
@ecorm I remember you said you're just using this thread as a convenient place for tracking the language standard adaptation, but I will use your posting as an opportunity to write down the remaining stuffs that need to be done😃

- Provide a CMake option for switching between language standards.
- Update test codes that remain as C++17+ only. I don't plan to update all test/meta codes to be C++11-compatible, but some needs to be.
- Update CI to test various standards.

Once these are done I will close this issue. Hopefully in the near future!

**Comment by ecorm** [comment]
@jk-jeon I'm now adopting C++17 as the minimum version for my libraries, but float/string conversion library is far more fundamental than the niche stuff I'm doing.

**Comment by jk-jeon** [comment]
Another remaining issue: there is a variable template (C++14) in `dragonbox_to_chars.h`. I need to detect existence of this feature and hide it if the feature doesn't exist. Right now, cppreference is down which has damaged my productivity to death, but maybe I can just read up the standard draft by myself.

>I use 23/26 standard in production only because.

Many of the commits after 1.1.3 tag were about experiments on how compact this library can be to be a good fit for embedded platforms, particularly AVR 8-bit. Turns out, many of embedded folks are stuck with GCC 7.x (gosh!), which does support C++17 but the support is marked experimental. So the effort of porting to C++11 wasn't a waste of time maybe? Unfortunately these ancient compilers seem to have hard time optimizing multiplications and shifts, so they produce much bigger binaries compared to GCC 15.x... which means I'm not sure if these folks will ever pick up this library. But anyway.

**Comment by vinniefalco** [comment]
You should ask chatgpt instead. I quickly got this answer, although I am not sure if it is correct https://chatgpt.com/share/68c3190e-ab64-800a-b446-76ec8a5c1007

**Comment by jk-jeon** [comment]
@vinniefalco Yeah, this sounds like a kind of tasks that LLM's should work decently. Thanks for reminding me!

**Comment by vinniefalco** [comment]
I'm still learning myself !

**Comment by ecorm** [comment]
https://devdocs.io hosts the entire cppreference and features an offline mode too. Many other languages and libraries on there too.

**Comment by ecorm** [comment]
Here are the results for "What version(s) of C++ are you allowed to use..." in [ISO's 2025 Annual C++ Developer Survey "Lite" (pdf)](https://isocpp.org/files/papers/CppDevSurvey-2025-summary.pdf):

<img width="1588" height="1368" alt="Image" src="https://github.com/user-attachments/assets/58b3883c-21ca-4a00-8abc-f4006088ff3a" />

**Comment by vinniefalco** [comment]
That's a very interesting statistic

**Comment by ecorm** [comment]
> Another remaining issue: there is a variable template (C++14) in `dragonbox_to_chars.h`. I need to detect existence of this feature and hide it if the feature doesn't exist.

FYI, a `constexpr` template function that returns the value would be the C++11-compatible way to do it instead of variable template.

> Many of the commits after 1.1.3 tag were about experiments on how compact this library can be to be a good fit for embedded platforms, particularly AVR 8-bit.

Why oh why are folks still using 8-bit micros when basic 32-bit micros cost next to nothing? Arduino is artificially keeping 8-bit AVRs alive when we should have all moved on to 32-bit.

**Comment by jk-jeon** [comment]
@ecorm Thanks a lot for useful info!

>FYI, a constexpr template function that returns the value would be the C++11-compatible way to do it instead of variable template.

Yeah but I don't want to break the API and it kind of feels weird to have both variable and function that do the same. So I just decided to wrap the variable inside a template struct (the classic C++03/11-era trick!) and have the outside variable template only for C++14+.

>Why oh why are folks still using 8-bit micros when basic 32-bit micros cost next to nothing? Arduino is artificially keeping 8-bit AVRs alive when we should have all moved on to 32-bit.

It was indeed quite amusing to see all the circus GCC is doing for producing binaries out of the standard C++ for these friendly little 8-bit guys.

I don't know if AVR 8-bit is used a lot for actual production, or it's just used mostly for education and prototyping. Do you know which CPU's are more commonly used for actual embedded products? I guess that's an overly general question given how diverse the embedded world is, but it would be helpful for me to hear some perspectives.

@vinniefalco 
>Consider reposting that here: boostorg/spirit_x4#4

I read the discussion about this in Slack. I guess... it's an interesting viewpoint on the C++ ecosystem.

**Comment by ecorm** [comment]
> Do you know which CPU's are more commonly used for actual embedded products? I guess that's an overly general question given how diverse the embedded world is, but it would be helpful for me to hear some perspectives.

I don't really know the stats, but I'm guessing the main ones are ARM and RISC-V-based. I like the ones where you're not forced to use the silicon vendor's proprietary IDE/compiler (which are often limited to Windows as the host platform). STM32 micros allow you to use VS Code with the GCC cross compiler. Nordic semiconductor is also good in that respect (they do MCU+radio chips). It seems the European chip makers are cluing in to idea that we like same familiar, powerful tool chains as for desktop/server projects.  I wish the rest of them would get the memo.

**Comment by jk-jeon** [comment]
@ecorm Thanks for the reply. Your reply led me to read many interesting write ups from various places.

**Comment by jk-jeon** [comment]
@vinniefalco I'm thinking of improving the CMake files, but I know very little of CMake so I asked ChatGPT to teach me. So far it seems working pretty good.

**Comment by saki7** [comment]
> Consider reposting that here: https://github.com/boostorg/spirit_x4/issues/4

I can't believe this comment was posted by Vinnie Falco.

@vinniefalco Vinnie, could you refrain from posting Spirit issues to irrelevant libraries? I did a quick full-text search on this repo and it seems like dragonbox is not depending on Spirit in the first place. I am so disappointed to see a Boost Organization member is creating emotional conflicts. Spirit and dragonbox are two different products and we have our own requirements. We already have had the conversation in our mailing list, and I thought that you apologized me for introducing controversial drama.

**Comment by vinniefalco** [comment]
Yes you are right and I have edited the post

**Comment by jk-jeon** [comment]
@saki7 I sincerely apologize if my comment sounded to you negatively. I read all the discussion from Spirit repo and Boost ML just purely out of curiosity and can sympathize your and other people's frustration. As you pointed out, this lib has nothing to do with Spirit, and I have absolutely no hard feeling about your work.

I hope you keep your hard work, and hope such an unfortunate communication problem does not occur again.

**Comment by alugowski** [comment]
> Another remaining issue: there is a variable template (C++14) in `dragonbox_to_chars.h`. I need to detect existence of this feature and hide it if the feature doesn't exist. 

>  GCC 7.x (gosh!), which does support C++17 but the support is marked experimental.

Variable template support appears to be added in GCC 5:

<img width="797" height="555" alt="Image" src="https://github.com/user-attachments/assets/79d0e820-7e27-41d7-acd6-534d3bc90420" />

So is there actually a need to do anything?

**Comment by jk-jeon** [comment]
@alugowski Formally speaking the feature was introduced in C++14 and the goal was to support C++11. I'm not sure how much practical value is in strictly following the standard language feature set, given that compilers that don't support anything beyond C++11 are likely buggy with C++11 features too and thus may not successfully compile this library anyway. But in any case I think it's a matter of promise and communication. If I say "it works with C++11" and if it doesn't work with some buggy C++11 compiler, then it's justified to say "it's their problem, not mine". Or we can look for workaround if that's feasible, but even for that case, simply knowing and agreeing upon that it's a workaround certainly matters, I think.

In any case, the needed work has in fact already been done yesterday.

**Comment by alugowski** [comment]
Makes sense.

Does the Dragonbox CI run the test suite on a variety of compilers? I do see the GitHub Actions workflow that calls `xcross` but I'm not familiar enough with that tool to tell if it only changes the platform or if it also lets you try different compilers.

If not then it's relatively straight forward to set that up with GitHub Actions. Install oldest GCC, clang, emscripten, MSVC, etc., also newest, and run the test suite in all c++ standards on each. It's also important to run on actual Windows and ARM macos hosts, not just virtualized on Linux hosts. This sounds excessive until it starts catching weird stuff.

**Comment by jk-jeon** [comment]
@alugowski That `xcross` stuff is contributed by its author and honestly I don't know a lot about it. I'm guessing that for each platform there is a compiler attached to it, and probably we don't have options for changing it to something else. I guess there may be options to specify the language version but I have to ask the author.

>If not then it's relatively straight forward to set that up with GitHub Actions. Install oldest GCC, clang, emscripten, MSVC, etc., also newest, and run the test suite in all c++ standards on each. It's also important to run on actual Windows and ARM macos hosts, not just virtualized on Linux hosts. This sounds excessive until it starts catching weird stuff.

This is what I was thinking. The `os.yml` one is doing the tests on Linux/MacOS/Windows (btw this was contributed by the same person), so maybe we just need to add language standards into it?

EDIT:
I missed the ARM part.

**Comment by alugowski** [comment]
Ah great. `macos-latest` does run on ARM these days. Windows ARM runners are coming AFAIK.

Yeah the `os.yaml` workflow would be easy to expand. Here is something I made about 2 years ago that would do it. https://github.com/alugowski/task-thread-pool/blob/main/.github/workflows/tests.yml . This one is a bit more complicated but also does MinGW and emscripten: https://github.com/alugowski/poolSTL/blob/main/.github/workflows/tests.yml

It uses third-party actions that install arbitrary versions of GCC and Clang. So all you need is a matrix of the ones you want. I haven't found one for MSVC, so my tests just use the default there.

In my opinion if Dragonbox works in C++11 mode on old GCC, old Clang, and whatever MSVC you're able to get, then you can declare this issue completed.

**Comment by jk-jeon** [comment]
@alugowski Yeah, there is a bit of caveat that I don't want to rewrite some of the test code to be C++11-compatible. I need to recap what are they and why I thought it's not worth doing so. Those tests need to be excluded from C++11 CI's.

**Comment by alugowski** [comment]
> [@alugowski](https://github.com/alugowski) Yeah, there is a bit of caveat that I don't want to rewrite some of the test code to be C++11-compatible. I need to recap what are they and why I thought it's not worth doing so. Those tests need to be excluded from C++11 CI's.

Yeah, though it seems you rolled your own test runner so you have some flexibility. Algorithmic tests can be in c++17 as long as the code is the same as if it would be in C++11 mode. This is a semi-common problem as all the modern test runners dropped C++11 support long ago. Google Test requires C++17 and Catch2 requires C++14 though you can use a very old branch to get C++11.

**Comment by vinniefalco** [comment]
Looking forward to this, as it will give Boost.JSON a speed increase :)

**Comment by IRainman** [comment]
[fmt](https://github.com/fmtlib/fmt) supports C++11. Why did you want more?

**Comment by IRainman** [comment]
Furthermore, you can use [fast_float](https://github.com/fastfloat/fast_float) for another way of conversion; it also supports C++11

**Comment by IRainman** [comment]
@vinniefalco it's for you. ;) fast_float also supports JSON format.
